File: leetcode/company-wise/microsoft/0340_longest_substring_k_distinct.py
# LC 340. Longest Substring with At Most K Distinct Characters | Medium | Microsoft
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s, k):
        from collections import defaultdict
        count = defaultdict(int); l = res = 0
        for r in range(len(s)):
            count[s[r]] += 1
            while len(count) > k:
                count[s[l]] -= 1
                if count[s[l]] == 0: del count[s[l]]
                l += 1
            res = max(res, r - l + 1)
        return res

File: leetcode/company-wise/microsoft/0340_longest_substring_k_distinct.cpp
// LC 340. Longest Substring with At Most K Distinct Characters | Medium | Microsoft
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        unordered_map<char, int> cnt; int l = 0, res = 0;
        for (int r = 0; r < s.size(); r++) {
            cnt[s[r]]++;
            while (cnt.size() > k) { if (--cnt[s[l]] == 0) cnt.erase(s[l]); l++; }
            res = max(res, r - l + 1);
        }
        return res;
    }
};
